File: src/main/java/org/noxet/noxetserver/util/RegionBinder.java
package org.noxet.noxetserver.util;

import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.noxet.noxetserver.messaging.Message;

import java.util.Set;

public class RegionBinder extends DynamicTimer {
    private final Location center;
    private final Set<Player> playerSet;
    private final int width, tickFrequency;

    public RegionBinder(Location center, Set<Player> playerSet, int width, int tickFrequency) {
        this.center = center;
        this.playerSet = playerSet;
        this.width = width;
        this.tickFrequency = tickFrequency;

        touchTimer();
    }

    @Override
    public boolean isTimerNecessary() {
        return !playerSet.isEmpty();
    }

    @Override
    public int getTickFrequency() {
        return tickFrequency;
    }

    @Override
    public void timerCall() {
        for(Player player : playerSet) {
            Location mobile = player.getLocation();
            Location delta = center.clone().subtract(mobile);

            if(Math.abs(delta.getX()) > width) {
                mobile.setX(center.getX() + width * -Math.signum(delta.getX()));
            } else if(Math.abs(delta.getZ()) > width) {
                mobile.setZ(center.getZ() + width * -Math.signum(delta.getZ()));
            } else continue;

            player.teleport(mobile);
            new Message("§cNuh uh!").send(player);
        }
    }
}

File: src/main/java/org/noxet/noxetserver/util/RegionBinder.scala
package org.noxet.noxetserver.util

import org.bukkit.Location
import org.bukkit.entity.Player
import org.noxet.noxetserver.messaging.Message

import scala.collection.immutable.HashSet

class RegionBinder(center: Location, playerSet: HashSet[Player], width: Int, tickFrequency: Int) extends DynamicTimer:
    touchTimer()

    override def isTimerNecessary: Boolean = playerSet.nonEmpty

    override def getTickFrequency: Int = tickFrequency

    override def timerCall(): Unit =
        for
            player <- playerSet
            mobile = player.getLocation
            delta = center.clone.subtract(mobile)
            if Math.abs(delta.getX) > width || Math.abs(delta.getZ) > width
        do
            mobile.setX(center.getX + width * -Math.signum(delta.getX))
            mobile.setZ(center.getZ + width * -Math.signum(delta.getZ))
            player.teleport(mobile)
            Message("§cNuh uh!").send(player)